### PR TITLE
LF-4164 Create animal details container

### PR DESCRIPTION
**Pull request by kathyavini**
**PLEASE READ THIS FIRST!!**

This is a very large PR 🫤 And as the many parts work interdependently, it was not possible to break it up and still allow the form as as a whole to be tested.  Therefore I have created a review guide that describes the changes within each section of code. _Please refer to the guide when looking at the PR, thank you!_

### Package Root
```
webapp/
├── package.json
└── pnpm-lock.yaml
```

- types were added for two libraries we already use

### Translations
```
webapp/public/locales
```

- self-explanatory

### RTK Query Configuration
```
src/
├── apiConfig.js
└── store/api/
    ├── apiSlice.ts
    └── types/
```
- self-explanatory

### Animal Form Components (Basics + Details)
```
Animals/
├── AddAnimalDetails/
├── AddAnimalsFormCard/ (= basics component)
  ├── AddAnimalsFormCard.tsx
  └── AnimalSelect.tsx
├── AddBatchDetails/
└── DetailCards/
  ├── General.tsx
  ├── Origin.tsx
  ├── Other.tsx
  └── Unique.tsx
```

- AddAnimalDetails and AddBatchDetails lay out the four subcomponent cards (General, Origin, Other, Unique), which are now in their own directory as they will also be used in the upcoming readonly and edit views
  - name prefixing has been added to all components to make them function within their respective field arrays
  - types have been moved from component to container so that the basics and details fields are defined together in the same file
  - some styling adjusts, plus moving the styles belonging to the subcomponents into that folder
  - some watch-requiring logic which controls the rendering of extra fields (`isOtherUseSelected`, `origin`, `shouldShowTagTypeInput`) that were set up as props are now moved into their respective components and the props removed

- AddAnimalsFormCard is the basics component (i.e. holds the form inputs for the Basics step). Changes:
  - constant which holds the form field names was renamed to BasicFields to match the convention of DetailsFields
  - form validation logic was added where missing (count, type, sexDetails)
  - added logic to populate the inputs according to form value (i.e. `value` prop from Controller render function)
  - card now generates and tracks a uuid ref (in a form field) which is used to sync each basics card with its details view entries. (Note: there is a useFieldArray-generated uuid called `field.id` that I could not use as an identifier for these purposes because it is not preserved by react-hook-form when navigating to the details step and back)
  - prevAnimalTypeRef was implemented because the useEffect that clears breed was triggering when navigating back to the basics step from details
  - The type AnimalSelect needed to pull and display its own error message
  
### Other Component Changes (minor)
```
components/
├── Expandable/
└── Form/
  ├── NumberInput/
  ├── SexDetails/
  └── hookformValidationUtils.js
```
- `<ExpandableItem />` was given a few more styling classes to alter wrapper styling based on expanded state
- `<NumberInput />` initialValue as given had been overriding form value
- minor type fix for sexDetails
- validation utils error messages were slightly logically incorrect

### Animal Containers
**THIS IS THE MAIN CONTENT OF THE PR!**
```
containers/Animals/AddAnimals/
├── AddAnimalBasics/
  ├── index.tsx
  └── useUpdateBasics.ts
├── AddAnimalDetails/
  ├── index.tsx
  └── usePopulateDetails.ts
├── types.ts
└── useAnimalOptions.ts
```

- Basics container
  - Refactored to use the new hook `useAnimalOptions.ts` which fetches and formats the React Select options for both containers
  - Has a new hook `useUpdateBasics.ts` which holds the logic for updating the basics form based on changes that were applied to the details form state when navigating back
  - `onRemove()` function now removes both its own field array entry and the corresponding part of the details field array
- Details container
  - Has its own hook `usePopulateDetails.ts` which holds the logic for updating the details form state based on the basic cards when navigating forward (Note: at the end of the PR description is a list of synchronization features that were implemented)
  - The remainder of the file formats the form data (= the useFieldArray DETAILS fields) to populate the AddAnimalDetails and AddBasicDetails components. I would particularly welcome suggestions for cleaning this up 🙏 

### Miscellaneous (Minor)
```
src/
├── containers/
  ├── Inventory/
    ├── useAnimalInventory.tsx
├── stories/
└── util/
    ├──animal.ts
```
- exported the function `getDefaultAnimalIconName` which MAY only be necessary for fixing the `PIG` vs `PIGS` situation (how did we mess that up? 😂) but might as well re-use written code for this. Icon behaviour is similar between here and Inventory
- _no significant changes to the Storybook stories_. Mostly renamed imports and types
  - name prefixing did at one point break the Storybook story logic
    - UPDATE: IT WAS THE DOT! Thank you @SayakaOno!! In the future we just want to use the name prefixing strategy used here (including the trailing `.` within the prefix + defaulting the prefix to an empty string) to let the components function inside or outside of field arrays
  - Since the watch-requiring logic of `isOtherUseSelected`, `origin`,  and `shouldShowTagTypeInput` were moved into their individual components, this logic could also be removed from the stories.
  - Added the same `KEY` prop now added by `useAnimalOptions` to the options mockData
- re-added a slightly different (default only) parse function to reverse the `'custom_4'` type ids we needed for the react select

That's it! Thank you! 🙇 

Jira link: https://lite-farm.atlassian.net/browse/LF-4164

**Further notes (for reference):**
Synchronization features implemented between the basics and details form steps:
- Forward propagation of basics data into the details field aray
- Array position-based preservation of details data (within each basic card group) when going to basics and back
- Forward propagation of delete when removing entire cards from basics
- Forward propagation of sexDetails -> sex radio options when creating individual animals
- Forward propagation of count (addition / deletion) to animal detail items
- Backwards propagation of full deletion via animal remove, i.e. if all the detail entries corresponding a basic card's data are removed, the basics card is deleted as well
- Backwards propagation of partial deletion via animal remove (count + sexDetails updated)
- Backwards propagation of sex -> sexDetails
- (Batches only) Synching of common form fields between forms (backwards propagation of batch name, count, and sexDetails)

Not implemented:
- Sex-based details identity. Details fields are synched to their specific basics card, but within that they are _only position-based_ so changing the sexDetails distribution will apply the new sexes in order. This means the 3rd animal might change its sex while maintaining its other properties like name, use, etc. I think sex-based identity is probably more intuitive, but it would be more complex than the position-based preservation of data we discussed last week with Loïc. We could revisit if desired.

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [ ] Passes test case
- [x] UI components visually reviewed on desktop view
- [x] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [x] I have added or updated language tags for text that's part of the UI
- [x] I have added "MISSING" for all new language tags to languages I don't speak
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

**Review by SayakaOno — COMMENTED** [review]
Amazing job! I have 4 more files to review, but I will submit what I've reviewed so far.
I think this PR is much smaller than you might think! The code is very readable, so it's not a pain at all :) Thank you for the guid as well, it was definitely helpful!

File: packages/webapp/src/components/Animals/DetailCards/General.tsx
Comment (kathyavini) [reply packages/webapp/src/components/Animals/DetailCards/General.tsx]: Thank you!

File: packages/webapp/src/store/api/types/index.ts
Comment (kathyavini) [reply packages/webapp/src/store/api/types/index.ts]: Oh dear, I think what I did was see that the type wasn't what I expected, but then changed the wrong `number` to `string`?? I wish that would have generated a type error, but I guess importing the options from the `useAnimalOptions` hook ruins that, maybe? 😞 

As for why the `default_type_id` was a string... that I didn't question, I just went with it! 😂

File: packages/webapp/src/containers/Animals/AddAnimals/types.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/types.ts:51]: Oh sorry, I did not compare any of these strings to the backend values; just kept them as the component types file had defined them.

File: packages/webapp/src/containers/Animals/AddAnimals/types.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/types.ts:1]: Thank you!!

File: packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx
Comment (kathyavini) [reply packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx]: Ah thank you! I think it was for the trigger, but then I didn't even use it there.

File: packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx
Comment (kathyavini) [reply packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx]: I think so, but I only added it to fix the timing of the `Required` error (so that it appears as soon as the input is cleared and resolves when a type is selected) so I'm not sure the identity of the type has been relevant yet... is there a reason you ask? Do you think it would be better in the Controller `field.onChange`?

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts]: Wait, you can index into a type???! 🤯 

I think in general I prefer naming the type for the non-array version, but this is a handy trick! I will use it here and for similar, thank you.

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts]: 😳

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts:37]: Thank you for spotting that!! 🙏  I think the bug is fixed with the dependency array change but I'm still a bit undecided about where the absolute best place to run the function would be.

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx:100]: Yes!

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (SayakaOno) [inline packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx:102]: 😭 It was such a bad idea to have the container hold the logic!!! I think I wasn't sure what the API responses for uses, sexes or origin would be like, so I went with an approach that wouldn't be affected by the API response structure 😞 Is it an option to move the components? 😢 

I didn't test it, but my approach for this component would be:
1. Create a function to generate `ExpandableItem` depending on `animal_or_batch` (almost like the callback function of the forEach).
2. Separate fields into animals and batches.
3. Render animals and batches like this?:
```
{animals.map((animal, index) => generateExpandableItem(animal, index))}
{batches.map((batch, index) => generateExpandableItem(batch, index))}
```

I think it would be easier to handle `animalIndex` and `batchIndex`.

Apart from that, I think `field.animal_or_batch === AnimalOrBatchKeys.ANIMAL` can be replaced with `isAnimal`  :)

File: packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx
Comment (SayakaOno) [reply packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx]: I commented without looking into `AnimalTypeSelect`, but I understand it now!
Initially, I wondered if the trigger should be called after `onTypeChange`, since I thought `onTypeChange` was responsible for setting the value (even though it shouldn't be, since it's optional). It's all good now!!

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts
Comment (SayakaOno) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts]: > I think in general I prefer naming the type for the non-array version
Me too!!!

File: packages/webapp/src/store/api/types/index.ts
Comment (SayakaOno) [reply packages/webapp/src/store/api/types/index.ts]: Okay, I found the cause. (I used `default_type_id` as map keys, and it became string.)
I will fix the API and adjust the frontend code after this PR is merged!

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx:100]: I will most happily move the logic into the components! 😄🙏 

Thank you for the render suggestion. That manual index increment felt like such a code smell but it was a long function and hard for me to think about. I'll take a fresh look after the other code cleanup is completed and try to implement what you describe above ❤️

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts]: Oh my goodness, thank you!! 100x better!

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx]: @SayakaOno I did exactly what you said (made the callback of the `forEach` a separate function and mapped over in the render itself) and I already think it's SOOO much better!! But could you take a look to see if it's implemented correctly according to what you were thinking? 🙏

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts
Comment (Duncan-Brain) [inline packages/webapp/src/containers/Animals/AddAnimals/AddAnimalBasics/useUpdateBasics.ts:34]: Would it make sense to pass this in from parent -- the parent calls this twice otherwise.

**Review by Duncan-Brain — APPROVED** [review]
Great job 👏  looks like it was a tough one to logic out even though it all looks well organized! I am seeing everything work I think as expected -- my questions are just code decisions questions I think! 

1. PIG vs PIGS lol!
- possible that function was there for Chicken --> Broilers + Layers and when I fixed I didn't look to simplify functions using it.
- lets just fix the IconKey in the mapping to PIGS! (its PIGS with 's' in DB -- I actually messed up and should have made Chicken plural) Then this can be simplified to just use the key directly or default.

2. useAnimalOptions()
- cool design and I think it makes sense -- this caught my eye because of the constants being used as props to the custom hook and made me wonder if avoiding mappings is really big efficiency saver vs. dev usability to have it just create everything without configuring props
- can consider using Anto's `useQueries()` hook but not a big deal

3. useUpdateBasics()
- Would it make sense to pass the sexDetails from parent instead of calling the `useAnimalOptions()` in this hook as well as the parent that calls this hook?

4. Specific situation I think is intended behaviour
-  if you make a basics card that creates individual animals with count: 3 and specify (1 male, 2 female). In next page, enter in female details first, then go back to Basics and remove the male and reduce count to 2. Then return to details -- it is expected the last female is removed not the "male" right -- the last ones are are deleted first in order of unspecified, female, male -- regardless of specified sexes.

5. MultiStepForm architecture + useUpdateBasics + usePopulateDetails
- Do you see any value in generalizing the custom hooks as something that will be useful as we continue using multistep form? 
  - the common structure I am seeing is along the lines of `useNextStepValues()` and `usePreviousStepValues()` 
      - definitions of the fields structure like create/update Animals
      - translation/mapping from stepA to stepB (or vice versa) fields
      - replacing/ updating the fields
      - caveat that this is currently only for many to many field arrays, and in the future would want to include non field array values
      
6. Not really related to this PR
- is `AddAnimalsDetails` component and `AddBatchDetails` component really just the same thing?

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/usePopulateDetails.ts
Comment (Duncan-Brain) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/usePopulateDetails.ts]: I kind of like it .. reminds me slightly of class structure and instantiating a new class!

Comment (SayakaOno) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/usePopulateDetails.ts]: > it doesn't add much
Instead, it initially added a few lines for `createAnimal`'s arguments, which I didn't love 😂 
I think I'm not used to see functions sitting inside `useEffect`...? But it looks like a good practice, so I'll try to do the same!

> I kind of like it .. reminds me slightly of class structure and instantiating a new class!
True 😄

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (SayakaOno) [inline packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx:88]: Now that you created commonProps, I feel like these could directly go into commonProps. You can leave it as is, I'm just obsessed with reducing lines 😂

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (SayakaOno) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx]: Thank you for moving the logic, it's much cleaner 😭 
`fieldArrayIndex` is something that I didn't take into account, thank you for handling it! Other than that, the new code is exactly what I was picturing. Thank you!!

File: packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx
Comment (SayakaOno) [inline packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx:150]: I don't think it's worth spending the time, but having the same logic in validate and `SexDetails` doesn't feel right 😓

**Comment by kathyavini** [comment]
Sorry for dismissing reviews but I am finding a few fresh bugs after the vigorous refactors 😂 Will re-request review when resolved, thank you 🙏

**Review by SayakaOno — COMMENTED** [review]
Thank you for moving the logic and adjusting stories!!

I just saw your comment @kathyavini, but I had recorded so will attach it 😂 

https://github.com/user-attachments/assets/94043a04-1312-47bf-803f-70223cc07a92

**Comment by kathyavini** [comment]
Thank you @SayakaOno! That wasn't even the bug I was thinking of, but that was an important one and an easy fix!

Unfortunately there is a worse bug 😞 I think it predated the refactors and as it's not that hard to re-create, I should have caught it earlier!

1. Create two animals or batches (doesn't matter)
2. Delete the first one
3. Return back + add a new card (two cards added) -OR- go directly forward (crash)

Despite only one card being rendered (correct), a `getValues()` on the form state shows an empty field with just a ref assigned:

Screenshot 2024-08-23 at 12 35 40 PM

I suspect the field array updating with the hook and the component setting its field array value are happening in a bit of a race, which is no good 😬 Also in that situation the valid card has switched its `field_array_id`, which seems dangerous on update. Still thinking on this one, and wondering if moving the update to the cleanup function could fix it...

**Comment by SayakaOno** [comment]
Right...! My idea won't work anyway because now the URLs don't change!
I was observing what's happening, your assumption is that `replace(filteredFields);` in `useUpdateBasics` isn't working correctly because rendering is happening simultaneously?

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx]: I wrote it that way too at first! But then I wondered if it was more efficient to define the props that don't change outside of the mapping callback function, so they are created only once? But I really wasn't sure -- it's probably a negligible difference, right??! I'll move them because I prefer not having to click through another definition to see what's in the prop 🙂

File: packages/webapp/src/containers/Animals/AddAnimals/useAnimalOptions.ts
Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/useAnimalOptions.ts:46]: As a rule, I don't use that hook because once you pass the queries through it, you lose the typing on the returned arrays. You could re-import the types manually but it kind of negates both the lines saved and the coolest part of RTK Query, which how the hooks kept track of the types defined in apliSlice (and there is always the possibility of introducing a bug when you're synching the types manually!)

The hook was also written in js before we shifted the frontend to (and wrote the apiSlice in) ts so we may just want to get rid of it.

File: packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx
Comment (kathyavini) [reply packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx:150]: I think I agreed with you on both counts here 😄 I initially tried passing the error as a prop after I added this validation, but when I saw that the message didn't triggering in the same way, I left it alone rather than looking any more deeply into how SexDetails was using `error`... it didn't seem worth the investigation at the time.

I think there is still a bit of room for refactor here though (like also maybe that extracted count + sexDetails component!) when we continue with the components in `readonly`/`edit`!

**Comment by kathyavini** [comment]
> My idea won't work anyway because now the URLs don't change!

@SayakaOno the URLs have to change for the component to be considered unmounted? Okay then I'm even more confused because it actually did trigger... sometimes 😅 It would be like on the 3rd iteration of going to details and back and then from then on!! It was so wacky!

>  your assumption is that replace(filteredFields); in useUpdateBasics isn't working correctly because rendering is happening simultaneously?

Yes, I think so. The hook does the filtered fields exactly right; it actually isn't affected by the simultaneous rendering since it's based on details, but the form state itself gets an extra `setValue()` applied. I assume the original fields map, call their `setValues()` and this happens at just about the same time the hook does the replace (or just after), so that bunk card is created with nothing but the field array id. I paused render with something like a loading state and I _think_ it's all good now! 🤞

@Duncan-Brain thanks for the comments and suggestions, I responded in-line to 2 & 3 and ignored 1 (the icons for now) as that will span a few parts of animals, we should do it at the end though!

(4) I think you're describing the same things as in the "Sex-based details identity" section at the end of the PR description, right? I think maybe we have to bring it to Loïc because I think everyone's intuition is going to be that sex controls the updating of animals within a given card when it is truly just index. If we don't address it, it might just keep coming up in QA and bug bash... 🫤 But it's a step beyond what we discussed with Loïc last week, which was saving details according to order.

(5) I initially thought this way, but I un-convinced myself just now when I was trying to implement the details -> basics synch within the cleanup function of details. The update functions need to access the `fields` and the useFieldArray `replace` specific to that array -- _but_ useFieldArray return methods aren't passed within `<FormContext />`. Do we want to create a new manual context in the multi-step form to hold all the array methods for each field array? As you said, field arrays and synching are a requirement of all forms either.

Personally, I think prefer the implemented pattern (on render, sync yourself) because it keeps everything local to the container. It also feels super intuitive, but since it's been getting push back from all fronts, maybe it's only intuitive to me 😂  I am open to discuss in tech daily!

(6) Hmm, yeah, could probably have been written with a prop. But I like how there isn't much internal logic required in them this way so I don't mind how it was set up.

File: packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx
Comment (SayakaOno) [inline packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx:154]: I'm hoping to avoid finding `fieldArrayIndex` in `generateExpandableItem`.

Is it safe to assume that the indices of the `fields` array correspond directly to `fieldArrayIndex`? If so, we could create `animals` and `batches` arrays while attaching `fieldArrayIndex`:

```
const animals = [] as (AnimalDetailsField & { fieldArrayIndex: number })[];
const batches = [] as (AnimalDetailsField & { fieldArrayIndex: number })[];

fields.forEach((animalOrBatch, index) => {
  const array = isAnimalField(animalOrBatch) ? animals : batches;
  array.push({ ...animalOrBatch, fieldArrayIndex: index });
});
```
Then, use it in the function: 

```
const generateExpandableItem = (
  fieldWithFieldArrayIndex: AnimalDetailsField & { fieldArrayIndex: number },
  index: number,
) => {
  const { fieldArrayIndex, ...field } = fieldWithFieldArrayIndex;
...
``` 
What do you think?

Comment (kathyavini) [reply packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx:154]: Well, it's not my first choice for the same reason I went with the double filter instead of `fields.forEach()` -- I usually prefer a little more readability over a little more efficiency for these small arrays -- but now I might be letting the inefficient things add up, so I'll use it 🙏 

(Also I like this a lot: `const array = isAnimalField(animalOrBatch) ? animals : batches` 😄)

**Review by SayakaOno — APPROVED** [review]
> > My idea won't work anyway because now the URLs don't change!
> 
> @SayakaOno the URLs have to change for the component to be considered unmounted? Okay then I'm even more confused because it actually did trigger... sometimes 😅 It would be like on the 3rd iteration of going to details and back and then from then on!! It was so wacky!

Sorry, my idea was to update the basic data only when leaving the details container and going back to the basic container, and I was thinking of doing something like this: `if (URL === basicsURL)`, which is impossible!


Thank you for providing your insights on the conflicting setValue!
Nice workaround, it's working great!

Great job on this huge task! ❤️❤️❤️

File: packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Animals/AddAnimalsFormCard/AddAnimalsFormCard.tsx:82]: Just some sort of form or container state or something.  uuid is good I just hadn't seen it used before and I think yours is easy to understand -- my first thought was is it a bulky import -- but I think its small.

**Review by Duncan-Brain — APPROVED** [review]
🎉  Nice looks good!

**Pull request merged**
